pkg/client/block/client.go

```diff
@@ -0,0 +1,169 @@
+) (client.BlockClient, error) {
+	// Initialize block client
+	bClient := &blockClient{endpointURL: cometWebsocketURL}
+	bClient.latestBlockObsvblsReplay, bClient.latestBlockObsvblsReplayPublishCh =
+		channel.NewReplayObservable[client.BlocksObservable](ctx, 1)
+
+	// Inject dependencies
+	if err := depinject.Inject(deps, &bClient.eventsClient); err != nil {
+		return nil, err
+	}
+
+	// Concurrently publish blocks to the observable emitted by latestBlockObsvblsReplay.
+	go bClient.goPublishBlocks(ctx)
+
+	return bClient, nil
+}
+
+// CommittedBlocksSequence returns a ReplayObservable, with a replay buffer size
+// of 1, which is notified when block commit events are received by the events
+// query subscription.
+func (bClient *blockClient) CommittedBlocksSequence(ctx context.Context) client.BlocksObservable {
+	replayedBlocksObservable := bClient.latestBlockObsvblsReplay.Last(ctx, 1)[0]
+	return replayedBlocksObservable
+}
+
+// LatestBlock returns the latest committed block that's been received by the
+// corresponding events query subscription.
+// It blocks until at least one block event has been received.
+func (bClient *blockClient) LatestBlock(ctx context.Context) (latestBlock client.Block) {
+	v := bClient.CommittedBlocksSequence(ctx).Last(ctx, 1)[0]
+	return v
+}
+
+// Close unsubscribes all observers of the committed blocks sequence observable
+// and closes the events query client.
+func (bClient *blockClient) Close() {
+	// Closing eventsClient will cascade unsubscribe and close downstream observers.
+	bClient.eventsClient.Close()
+}
+
+// goPublishBlocks receives event bytes from the events query client, maps them
+// to block events, and publishes them to the latestBlockObsvblsReplay replay observable.
+func (bClient *blockClient) goPublishBlocks(ctx context.Context) {
+	// NB: cometbft event subscription query
+	// (see: https://docs.cosmos.network/v0.47/core/events#subscribing-to-events)
+	query := "tm.event='NewBlock'"
+
+	// React to errors by getting a new events bytes observable, re-mapping it,
+	// and send it to latestBlockObsvblsReplayPublishCh such that
+	// latestBlockObsvblsReplay.Last(ctx, 1) will return it.
+	retryOnError(ctx, "goPublishBlocks", func() chan error {
+		errCh := make(chan error, 1)
+		eventsBzObsvbl, err := bClient.eventsClient.EventsBytes(ctx, query)
+		if err != nil {
+			errCh <- err
+			return errCh
+		}
+
+		// NB: must cast back to generic observable type to use with Map.
+		// client.BlocksObservable is only used to workaround gomock's lack of
+		// support for generic types.
+		eventsBz := observable.Observable[either.Either[[]byte]](eventsBzObsvbl)
+		blocksObsvbl := channel.MapReplay(ctx, 1, eventsBz, blockEventFromEventBz)
+
+		// Initially set latestBlockObsvblsReplay and update if after retrying on error.
+		bClient.latestBlockObsvblsReplayPublishCh <- blocksObsvbl
+
+		return errCh
+	})
+}
+
+// retryOnError runs the given function, which is expected to return an error
```

Comment:
```suggestion
// retryOnError runs the given function, which could return an error
```

Reply:
Is that correct or no?

Reply:
The relevant sentence fragment is "is expected to return an error **channel**"; however, "channel" was wrapped to the next line. I've reformatted to mitigate this confusion going forward as well as adding more detail and adding a branch for `retryOnError` to exit (i.e. instead of run forever :sweat_smile:).

Reply:
I moved `retryOnError` to its own pkg and added some tests.